File: adapters/python/synapticrelay/manifest.py
"""
Manifest builder and validator for Python agent runtimes.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def validate_manifest(manifest: dict[str, Any]) -> tuple[bool, list[str]]:
    """Validate a manifest against basic spec requirements.

    Returns:
        Tuple of (is_valid, list_of_errors)
    """
    errors: list[str] = []

    # Required top-level fields
    if manifest.get("specVersion") != "1.0":
        errors.append("specVersion must be '1.0'")

    runtime = manifest.get("runtime")
    if not isinstance(runtime, dict):
        errors.append("runtime object is required")
    else:
        if not runtime.get("name"):
            errors.append("runtime.name is required")
        if runtime.get("type") not in ("openclaw", "python", "node", "mcp", "http", "custom"):
            errors.append("runtime.type must be one of: openclaw, python, node, mcp, http, custom")
        if not runtime.get("version"):
            errors.append("runtime.version is required")

    role = manifest.get("role")
    if role not in ("supplier", "buyer", "both"):
        errors.append("role must be one of: supplier, buyer, both")

    endpoints = manifest.get("endpoints")
    if not isinstance(endpoints, dict):
        errors.append("endpoints object is required")
    elif not endpoints.get("health"):
        errors.append("endpoints.health is required")

    # Semantic checks
    if role in ("supplier", "both"):
        caps = manifest.get("capabilities")
        if not caps or len(caps) == 0:
            errors.append("capabilities are required when role is 'supplier' or 'both'")
        if isinstance(endpoints, dict) and not endpoints.get("invoke"):
            errors.append("endpoints.invoke is required when role is 'supplier' or 'both'")

    return (len(errors) == 0, errors)


class ManifestBuilder:
    """Fluent builder for creating valid SynapticRelay manifests.

    Example:
        manifest = (
            ManifestBuilder("My Agent", "python", "1.0.0")
            .role("supplier")
            .description("A helpful agent")
            .health_endpoint("http://localhost:8000/health")
            .invoke_endpoint("http://localhost:8000/invoke")
            .add_capability("summarize", "Summarize documents")
            .build()
        )
    """

    def __init__(self, name: str, runtime_type: str, version: str):
        self._manifest: dict[str, Any] = {
            "specVersion": "1.0",
            "runtime": {
                "name": name,
                "type": runtime_type,
                "version": version,
            },
            "role": "supplier",
            "endpoints": {"health": ""},
        }

    def role(self, role: str) -> "ManifestBuilder":
        """Set the runtime role (supplier, buyer, both)."""
        self._manifest["role"] = role
        return self

    def description(self, desc: str) -> "ManifestBuilder":
        """Set runtime description."""
        self._manifest["runtime"]["description"] = desc
        return self

    def homepage(self, url: str) -> "ManifestBuilder":
        """Set runtime homepage URL."""
        self._manifest["runtime"]["homepage"] = url
        return self

    def health_endpoint(self, url: str) -> "ManifestBuilder":
        """Set the health check endpoint."""
        self._manifest["endpoints"]["health"] = url
        return self

    def invoke_endpoint(self, url: str) -> "ManifestBuilder":
        """Set the invocation endpoint."""
        self._manifest["endpoints"]["invoke"] = url
        return self

    def status_endpoint(self, url: str) -> "ManifestBuilder":
        """Set the status polling endpoint."""
        self._manifest["endpoints"]["status"] = url
        return self

    def webhook_endpoint(self, url: str) -> "ManifestBuilder":
        """Set the webhook callback endpoint."""
        self._manifest["endpoints"]["webhook"] = url
        return self

    def add_capability(
        self,
        name: str,
        description: str,
        input_schema: dict | None = None,
        output_schema: dict | None = None,
        category: str | None = None,
        tags: list[str] | None = None,
    ) -> "ManifestBuilder":
        """Add a capability to the manifest."""
        if "capabilities" not in self._manifest:
            self._manifest["capabilities"] = []

        cap: dict[str, Any] = {"name": name, "description": description}
        if input_schema:
            cap["inputSchema"] = input_schema
        if output_schema:
            cap["outputSchema"] = output_schema
        if category:
            cap["category"] = category
        if tags:
            cap["tags"] = tags

        self._manifest["capabilities"].append(cap)
        return self

    def invocation(
        self,
        mode: str = "sync",
        timeout_ms: int = 30000,
        max_concurrency: int = 10,
        retryable: bool = True,
    ) -> "ManifestBuilder":
        """Set invocation configuration."""
        self._manifest["invocation"] = {
            "mode": mode,
            "timeoutMs": timeout_ms,
            "maxConcurrency": max_concurrency,
            "retryable": retryable,
        }
        return self

    def auth(self, auth_type: str = "api_key", header_name: str = "X-API-Key") -> "ManifestBuilder":
        """Set auth configuration."""
        self._manifest["auth"] = {"type": auth_type, "headerName": header_name}
        return self

    def settlement(self, supported: bool = False, methods: list[str] | None = None) -> "ManifestBuilder":
        """Set settlement configuration."""
        data: dict[str, Any] = {"supported": supported}
        if methods:
            data["methods"] = methods
        self._manifest["settlement"] = data
        return self

    def metadata(self, **kwargs: Any) -> "ManifestBuilder":
        """Set arbitrary metadata."""
        self._manifest.setdefault("metadata", {}).update(kwargs)
        return self

    def build(self) -> dict[str, Any]:
        """Build and validate the manifest. Raises ValueError if invalid."""
        valid, errors = validate_manifest(self._manifest)
        if not valid:
            raise ValueError(f"Invalid manifest: {'; '.join(errors)}")
        return dict(self._manifest)

    def build_unchecked(self) -> dict[str, Any]:
        """Build without validation (useful during development)."""
        return dict(self._manifest)

    def to_json(self, indent: int = 2) -> str:
        """Serialize manifest to JSON string."""
        return json.dumps(self.build(), indent=indent)

    def to_file(self, path: str | Path) -> None:
        """Write manifest to a JSON file."""
        Path(path).write_text(self.to_json())

File: packages/core/src/manifest.ts
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import type {
  RuntimeManifest,
  Capability,
  RuntimeRole,
  RuntimeType,
  InvocationMode,
  AuthType,
  SettlementMethod,
  ValidationResult,
} from './types';
import { ManifestValidationError } from './errors';
import * as fs from 'fs';
import * as path from 'path';

// ─── Manifest Validator ─────────────────────────────────────────────

let cachedSchema: Record<string, unknown> | null = null;

function loadSchema(): Record<string, unknown> {
  if (cachedSchema) return cachedSchema;

  // Try multiple paths to find the schema
  const possiblePaths = [
    path.resolve(__dirname, '../../../spec/manifest.schema.json'),
    path.resolve(__dirname, '../../spec/manifest.schema.json'),
    path.resolve(process.cwd(), 'spec/manifest.schema.json'),
  ];

  for (const schemaPath of possiblePaths) {
    try {
      const raw = fs.readFileSync(schemaPath, 'utf-8');
      cachedSchema = JSON.parse(raw);
      return cachedSchema!;
    } catch {
      // Try next path
    }
  }

  throw new Error(
    'Could not find manifest.schema.json. Make sure you are running from the repo root or specify the schema path.',
  );
}

/**
 * Validate a manifest against the canonical JSON Schema.
 *
 * @returns ValidationResult with errors if invalid
 */
export function validateManifest(manifest: unknown): ValidationResult {
  const schema = loadSchema();
  const ajv = new Ajv({ allErrors: true, strict: false });

  // Add format validation support
  try {
    addFormats(ajv);
  } catch {
    // ajv-formats may not be installed; URI format checking is optional
  }

  const validate = ajv.compile(schema);
  const valid = validate(manifest);

  if (valid) {
    // Additional semantic validation
    const semanticErrors = validateSemantics(manifest as RuntimeManifest);
    if (semanticErrors.length > 0) {
      return { valid: false, errors: semanticErrors };
    }
    return { valid: true, errors: [] };
  }

  const errors = (validate.errors || []).map((err) => ({
    path: err.instancePath || '/',
    message: err.message || 'Unknown validation error',
    keyword: err.keyword,
  }));

  return { valid: false, errors };
}

/**
 * Validate a manifest and throw if invalid.
 */
export function assertManifestValid(manifest: unknown): asserts manifest is RuntimeManifest {
  const result = validateManifest(manifest);
  if (!result.valid) {
    throw new ManifestValidationError(
      `Manifest validation failed with ${result.errors.length} error(s)`,
      result.errors,
    );
  }
}

/**
 * Validate manifest from a JSON file path.
 */
export function validateManifestFile(filePath: string): ValidationResult {
  const raw = fs.readFileSync(filePath, 'utf-8');
  let manifest: unknown;
  try {
    manifest = JSON.parse(raw);
  } catch {
    return {
      valid: false,
      errors: [{ path: '/', message: `Invalid JSON: ${filePath}` }],
    };
  }
  return validateManifest(manifest);
}

// ─── Semantic Validation ────────────────────────────────────────────

function validateSemantics(manifest: RuntimeManifest): Array<{ path: string; message: string }> {
  const errors: Array<{ path: string; message: string }> = [];

  // Suppliers and both-role must have capabilities
  if ((manifest.role === 'supplier' || manifest.role === 'both') &&
      (!manifest.capabilities || manifest.capabilities.length === 0)) {
    errors.push({
      path: '/capabilities',
      message: 'Capabilities are required when role is "supplier" or "both"',
    });
  }

  // Suppliers and both-role must have invoke endpoint
  if ((manifest.role === 'supplier' || manifest.role === 'both') && !manifest.endpoints.invoke) {
    errors.push({
      path: '/endpoints/invoke',
      message: 'Invoke endpoint is required when role is "supplier" or "both"',
    });
  }

  // Capability names must be unique
  if (manifest.capabilities) {
    const names = new Set<string>();
    for (const cap of manifest.capabilities) {
      if (names.has(cap.name)) {
        errors.push({
          path: `/capabilities/${cap.name}`,
          message: `Duplicate capability name: "${cap.name}"`,
        });
      }
      names.add(cap.name);
    }
  }

  return errors;
}

// ─── Manifest Builder ───────────────────────────────────────────────

/**
 * Fluent builder for creating valid RuntimeManifest objects.
 *
 * @example
 * ```ts
 * const manifest = new ManifestBuilder('My Agent', 'python', '1.0.0')
 *   .role('supplier')
 *   .healthEndpoint('http://localhost:8000/health')
 *   .invokeEndpoint('http://localhost:8000/invoke')
 *   .addCapability({
 *     name: 'summarize',
 *     description: 'Summarize text documents',
 *   })
 *   .build();
 * ```
 */
export class ManifestBuilder {
  private manifest: RuntimeManifest;

  constructor(name: string, type: RuntimeType, version: string) {
    this.manifest = {
      specVersion: '1.0',
      runtime: { name, type, version },
      role: 'supplier',
      endpoints: { health: '' },
    };
  }

  /** Set the runtime role (supplier, buyer, both) */
  setRole(role: RuntimeRole): this {
    this.manifest.role = role;
    return this;
  }

  /** Set runtime description */
  description(desc: string): this {
    this.manifest.runtime.description = desc;
    return this;
  }

  /** Set runtime homepage URL */
  homepage(url: string): this {
    this.manifest.runtime.homepage = url;
    return this;
  }

  /** Set health endpoint URL */
  healthEndpoint(url: string): this {
    this.manifest.endpoints.health = url;
    return this;
  }

  /** Set invoke endpoint URL */
  invokeEndpoint(url: string): this {
    this.manifest.endpoints.invoke = url;
    return this;
  }

  /** Set status polling endpoint URL */
  statusEndpoint(url: string): this {
    this.manifest.endpoints.status = url;
    return this;
  }

  /** Set webhook callback endpoint URL */
  webhookEndpoint(url: string): this {
    this.manifest.endpoints.webhook = url;
    return this;
  }

  /** Add a capability */
  addCapability(cap: Capability): this {
    if (!this.manifest.capabilities) {
      this.manifest.capabilities = [];
    }
    this.manifest.capabilities.push(cap);
    return this;
  }

  /** Set invocation configuration */
  invocation(config: {
    mode?: InvocationMode;
    timeoutMs?: number;
    maxConcurrency?: number;
    retryable?: boolean;
  }): this {
    this.manifest.invocation = config;
    return this;
  }

  /** Set auth configuration */
  auth(config: { type?: AuthType; headerName?: string }): this {
    this.manifest.auth = config;
    return this;
  }

  /** Set settlement configuration */
  settlement(config: { supported?: boolean; methods?: SettlementMethod[] }): this {
    this.manifest.settlement = config;
    return this;
  }

  /** Set arbitrary metadata */
  meta(data: Record<string, unknown>): this {
    this.manifest.metadata = { ...this.manifest.metadata, ...data };
    return this;
  }

  /** Build and validate the manifest */
  build(): RuntimeManifest {
    assertManifestValid(this.manifest);
    return { ...this.manifest };
  }

  /** Build without validation (useful during development) */
  buildUnchecked(): RuntimeManifest {
    return { ...this.manifest };
  }
}
